rpc/handler.go

```diff
@@ -113,16 +184,35 @@ func (h *handler) handleBatch(msgs []*jsonrpcMessage) {
 	}
 	// Process calls on a goroutine because they may block indefinitely:
 	h.startCallProc(func(cp *callProc) {
-		answers := make([]*jsonrpcMessage, 0, len(msgs))
-		for _, msg := range calls {
-			if answer := h.handleCallMsg(cp, msg); answer != nil {
-				answers = append(answers, answer)
+		var (
+			timer       *time.Timer
+			callBuffer  = &batchCallBuffer{calls: calls, answers: make([]*jsonrpcMessage, 0, len(calls))}
+			ctx, cancel = context.WithCancel(cp.ctx)
+		)
+		cp.ctx = ctx
+		if deadline, ok := ctx.Deadline(); ok {
+			timer = time.AfterFunc(computeTimeout(deadline), func() {
+				cancel()
+				callBuffer.timeout(ctx, h.conn)
+			})
+		}
+		for {
+			// No need to handle rest of calls if timed out.
+			if ctx.Err() != nil {
+				break
+			}
+			msg := callBuffer.popCall()
+			if msg == nil {
+				break
 			}
+			answer := h.handleCallMsg(cp, msg)
+			callBuffer.pushAnswer(answer)
```

Comment:
For the timeouts, you take care and ensure that you don't push a response to a notification, but here you unconditionally push the anwer.  The previous code only did something with the answer after checking `answer != nil`, and I see you have moved that check into `handleCall`. 
I suspect that this works fine, but perhaps a more robust way to do it would be to 

```golang
if !msg.IsNotification() { 
    callBuffer.pushAnswer(answer)
}
```